File: backend-java/src/main/java/com/example/calorie/common/ApiException.java
package com.example.calorie.common;

import org.springframework.http.HttpStatus;

/**
 * クライアントへそのまま返してよい業務エラー。
 *
 * <p>スタックトレースを持たせないのは、業務エラーは想定内の分岐であり
 * 例外生成コストを払う意味がないため。
 */
public class ApiException extends RuntimeException {

    private final HttpStatus status;
    private final String code;

    public ApiException(HttpStatus status, String code, String message) {
        super(message, null, false, false);
        this.status = status;
        this.code = code;
    }

    public static ApiException badRequest(String code, String message) {
        return new ApiException(HttpStatus.BAD_REQUEST, code, message);
    }

    public static ApiException unauthorized(String code, String message) {
        return new ApiException(HttpStatus.UNAUTHORIZED, code, message);
    }

    public static ApiException conflict(String code, String message) {
        return new ApiException(HttpStatus.CONFLICT, code, message);
    }

    public static ApiException notFound(String code, String message) {
        return new ApiException(HttpStatus.NOT_FOUND, code, message);
    }

    public HttpStatus getStatus() {
        return status;
    }

    public String getCode() {
        return code;
    }
}

File: backend-kotlin/src/main/kotlin/com/example/calorie/common/ApiException.kt
package com.example.calorie.common

import org.springframework.http.HttpStatus

/**
 * クライアントへそのまま返してよい業務エラー。
 *
 * スタックトレースを持たせないのは、業務エラーは想定内の分岐であり
 * 例外生成コストを払う意味がないため。
 */
class ApiException(
    val status: HttpStatus,
    val code: String,
    override val message: String,
) : RuntimeException(message, null, false, false) {

    companion object {
        fun badRequest(code: String, message: String) =
            ApiException(HttpStatus.BAD_REQUEST, code, message)

        fun unauthorized(code: String, message: String) =
            ApiException(HttpStatus.UNAUTHORIZED, code, message)

        fun conflict(code: String, message: String) =
            ApiException(HttpStatus.CONFLICT, code, message)

        fun notFound(code: String, message: String) =
            ApiException(HttpStatus.NOT_FOUND, code, message)
    }
}
